full-service/src/service/t3_sync.rs

```diff
@@ -0,0 +1,152 @@
+    join_handle: Option<thread::JoinHandle<()>>,
+
+    /// Stop trigger, used to signal the thread to terminate.
+    stop_requested: Arc<AtomicBool>,
+}
+
+impl T3SyncThread {
+    pub fn start(config: T3Config, wallet_db: WalletDb, logger: Logger) -> Self {
+        let stop_requested = Arc::new(AtomicBool::new(false));
+        let thread_stop_requested = stop_requested.clone();
+
+        let t3_connection =
+            T3Connection::new(&config.t3_uri, config.t3_api_key.clone(), logger.clone());
+
+        let join_handle = Some(
+            thread::Builder::new()
+                .name("t3_sync".to_string())
+                .spawn(move || {
+                    log::debug!(logger, "T3 Sync thread started.");
+
+                    let conn = &mut wallet_db
+                        .get_pooled_conn()
+                        .expect("failed getting wallet db connection");
+
+                    loop {
+                        if thread_stop_requested.load(Ordering::SeqCst) {
+                            log::debug!(logger, "T3SyncThread stop requested.");
+                            break;
+                        }
+
+                        match sync_txos(conn, &t3_connection, &logger) {
+                            Ok(()) => (),
+                            Err(e) => log::error!(&logger, "Error during t3 sync:\n{:?}", e),
+                        }
+
+                        // This sleep is to allow other API calls that need access to the database a
+                        // chance to execute, because the t3 sync process requires a write lock on
+                        // the database.
+                        thread::sleep(std::time::Duration::from_millis(T3_SYNC_INTERVAL));
+                    }
+                    log::debug!(logger, "T3SyncThread stopped.");
+                })
+                .expect("failed starting t3 sync thread"),
+        );
+
+        Self {
+            join_handle,
+            stop_requested,
+        }
+    }
+
+    pub fn stop(&mut self) {
+        self.stop_requested.store(true, Ordering::SeqCst);
+        if let Some(join_handle) = self.join_handle.take() {
+            join_handle.join().expect("T3SyncThread join failed");
+        }
+    }
+}
+
+impl Drop for T3SyncThread {
+    fn drop(&mut self) {
+        self.stop();
+    }
+}
+
+pub fn sync_txos(
+    conn: Conn,
+    t3_connection: &T3Connection,
+    logger: &Logger,
+) -> Result<(), T3SyncError> {
+    // get all txos from the database that haven't synced yet to t3 and that have an
+    // authenticated sender memo
+    log::debug!(logger, "Syncing txos to t3");
+
+    let txos = Txo::get_txos_that_need_to_be_synced_to_t3(Some(TXO_CHUNK_SIZE), conn)?;
+
+    for txo in txos {
+        let txo_memo = txo.memo(conn)?;
+        let memo = match txo_memo {
+            TxoMemo::AuthenticatedSender(memo) => memo,
+            _ => continue,
+        };
+
+        sync_txo(&txo, &memo, t3_connection)?;
```

Comment:
How does t3 behaves if the transaction has already been synced? e.g. what happens if we sync the same tx twice for some reason?

Reply:
The `t3.CreateTransaction` RPC returns `RESULT_ALREADY_EXISTS`, which can be treated equivalently to `RESULT_OK`